### PR TITLE
Ensure that similar history requests won't be made using requestMessages

### PR DESCRIPTION
**Pull request by dshulyak**
This method will return error if multiple requests with same topics were made without any delay (by default 3 seconds, can be configured with ShhextConfig). With force=true in requests parameters described behavior will be ignored.

### CODE REVIEW

**Comment by ghost** [comment]
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

File: vendor/github.com/status-im/whisper/whisperv6/message.go [hunk 1]
```diff
@@ -75,6 +75,7 @@ type ReceivedMessage struct {
 
 	SymKeyHash   common.Hash // The Keccak256Hash of the key
 	EnvelopeHash common.Hash // Message envelope hash to act as a unique id
+	History      bool
```
Comment (mandrigin) [inline vendor/github.com/status-im/whisper/whisperv6/message.go:78]: Unnecessary change?

**Comment by mandrigin** [comment]
@dshulyak it would be nice to see a few tests of this method if possible.

File: vendor/github.com/status-im/whisper/whisperv6/message.go (hunk 1 above)
Comment (dshulyak) [reply vendor/github.com/status-im/whisper/whisperv6/message.go:78]: right, should not be here

File: cmd/statusd/flags_test.go [hunk 2]
```diff
@@ -83,7 +83,7 @@ func TestStatusFlag(t *testing.T) {
 		}
 
 		require.Nil(t, err, msg)
-		require.Equal(t, s.enabled, c.StatusServiceEnabled, msg)
```
Comment (mandrigin) [inline cmd/statusd/flags_test.go]: 👍 for the rename, it is much better now

**Comment by mandrigin** [comment]
@dshulyak one question: why is that a separate method instead of it being an optional parameter for the existing method?

**Review by mandrigin — APPROVED** [review]

**Comment by dshulyak** [comment]
> @dshulyak one question: why is that a separate method instead of it being an optional parameter for the existing method?

this is definitely an option. i don't have strong arguments for the separate method. it just seemed like a cleaner way to isolate this new feature. maybe that is not true.

**Comment by mandrigin** [comment]
@dshulyak just "safe" is a bit generic word. I would use a parameter, like `force: true` to have a single method.
Then normal `shhext_requestMessages` will be limited, where `shhext_requestMessages(force = true)` will ignore the limitation.

File: services/shhext/api.go [hunk 3]
```diff
@@ -178,7 +178,18 @@ func (api *PublicAPI) getPeer(rawurl string) (*enode.Node, error) {
...
+	return api.requestMessages(ctx, r, false)
+}
+
+// SafeRequestMessages provides same functionality as RequestMessages. But if similar request was made within
+// configured delay an error will be returned.
+func (api *PublicAPI) SafeRequestMessages(ctx context.Context, r MessagesRequest) (hexutil.Bytes, error) {
+	return api.requestMessages(ctx, r, true)
+}
+
+// RequestMessages sends a request for historic messages to a MailServer.
```
Comment (pedropombeiro) [inline services/shhext/api.go]:
```suggestion
// requestMessages sends a request for historic messages to a MailServer.
```

File: services/shhext/requests.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+// Register request with given topics. If request with same topics was made in less then configured delay then error
+// will be returned.
+func (r *RequestsRegistry) Register(uid common.Hash, topics []whisper.TopicType) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	topicsHash := topicsToHash(topics)
+	if meta, exist := r.byTopicsHash[topicsHash]; exist {
+		if time.Since(meta.timestamp) < r.delay {
+			return fmt.Errorf("another request with same topics was sent in less then %s ago. Please wait more time, or use shhext_requestMessages", r.delay)
```
Comment (pedropombeiro) [inline services/shhext/requests.go]:
```suggestion
			return fmt.Errorf("another request with the same topics was sent less than %s ago. Please wait for a bit longer, or use shhext_requestMessages", r.delay)
```

**Review by pedropombeiro — APPROVED** [review]
lgtm

**Comment by dshulyak** [comment]
variant with force looks better indeed. will apply

File: services/shhext/requests.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	topicsHash, exist := r.uidToTopics[uid]
+	if !exist {
+		return
+	}
+	delete(r.uidToTopics, uid)
+	meta := r.byTopicsHash[topicsHash]
+	// another request was made.
+	if meta.lastUID == uid {
```
Comment (adambabik) [inline services/shhext/requests.go:71]: I guess `meta.lastUID == uid` means that there were no other requests? If so then the comment is inaccurate.

File: services/shhext/requests.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+// topicsToHash returns non-cryptographic hash of the topics.
+func topicsToHash(topics []whisper.TopicType) common.Hash {
+	hash := fnv.New32()
+	for i := range topics {
+		hash.Write(topics[i][:]) // never returns error per documentation
+	}
+	rst := make([]byte, 0, 32)
+	rst = hash.Sum(rst)
```
Comment (adambabik) [inline services/shhext/requests.go]: Won't `rst := hash.Sum(nil)` produce the same result without a need to specify the slice explicitly? Questions arise when looking at this fragment like "why `32`?" and "what if `New32` changes to something else" etc.

**Review by adambabik — APPROVED** [review]

**Comment by adambabik** [comment]
> Then normal shhext_requestMessages will be limited, where shhext_requestMessages(force = true) will ignore the limitation.

This will introduce incompatibility, I believe. All existing calls to `shhext_requestMessages` will need to be fixed and send this additional param. It's fine but we need to remember about it and bump the `VERSION` properly.

**Comment by mandrigin** [comment]
@adambabik yes, it will introduce incompatibility, but that is what we want because the current behaviour forces it to use too much trafic.

**Comment by adambabik** [comment]
@mandrigin then it's cool :) wanted to make sure everyone is aware.

**Comment by mandrigin** [comment]
@adambabik yep, that is important to tell explicitly, thanks for pointing this out!

File: services/shhext/api.go [hunk 7]
```diff
@@ -178,7 +178,18 @@ func (api *PublicAPI) getPeer(rawurl string) (*enode.Node, error) {
...
+func (api *PublicAPI) requestMessages(_ context.Context, r MessagesRequest, ensureDelay bool) (hexutil.Bytes, error) {
```
Comment (mandrigin) [inline services/shhext/api.go]: hmm, isn't `ensureDelay` the opposite of `force`?

Comment (dshulyak) [reply services/shhext/api.go]: yes, but that code was without latest changes. with new changes i am using same method with force parameter

**Comment by mandrigin** [comment]
linter looks unhappy :(
